sale_coupon_advanced/models/sale_coupon_program.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+
+
+class SaleCouponProgram(models.Model):
+    _inherit = "sale.coupon.program"
+
+    is_cumulative = fields.Boolean(string="None-cumulative Promotion")
+    is_reward_product_forced = fields.Boolean(
+        string="Unordered product",
+        default=False,
+        help="If checked, the reward product will be added if not ordered.",
+    )
+
+    def _check_promo_code(self, order, coupon_code):
+        """ Do not return product unordered error message if
+            `is_reward_product_forced` is selected
+        """
+        message = _(
+            "The reward products should be in the sales order lines to"
+            " apply the discount."
+        )
+        res = super()._check_promo_code(order, coupon_code)
+        if res.get("error") == message and self.is_reward_product_forced:
+            return {}
+        return res
+
+    @api.model
+    def _filter_programs_from_common_rules(self, order, next_order=False):
```

Comment:
can we just force adding all the missed products to the order and relying rest on default behavior?
I would create a "new" sale order line with the product, call an onchange_product_id, and got a default price